precompiles/parachain-staking/src/lib.rs

```diff
@@ -232,286 +126,392 @@ where
 		Ok(PrecompileOutput {
 			exit_status: ExitSucceed::Returned,
-			cost: gas_consumed,
-			output: buffer.to_vec(),
-			logs: Default::default(),
+			cost: gasometer.used_gas(),
+			output: EvmDataWriter::new().write(is_candidate).build(),
+			logs: vec![],
 		})
 	}
 
-	fn points(input: &[u8]) -> Result<PrecompileOutput, ExitError> {
-		let round_u256 = parse_uint256(input)?;
+	fn is_selected_candidate(
+		mut input: EvmDataReader,
+		target_gas: Option<u64>,
+	) -> Result<PrecompileOutput, ExitError> {
+		let mut gasometer = Gasometer::new(target_gas);
 
-		// Make sure the round number fits in a u32
-		if round_u256.leading_zeros() < 256 - 32 {
-			return Err(error("Round is too large. 32 bit maximum"));
-		}
-		let round: u32 = round_u256.low_u32();
+		// Read input.
+		input.expect_arguments(1)?;
+		let raw_address: H160 = input.read::<Address>()?.into();
+		let address: Runtime::AccountId = raw_address.into();
 
-		log::trace!(target: "staking-precompile", "🥩round is {}", round);
-		// Read the point value and format it for Solidity
-		let points: u32 = parachain_staking::Pallet::<Runtime>::points(round);
-		log::trace!(target: "staking-precompile", "🥩points is {}", points);
-		let mut output = [0u8; 32];
-		U256::from(points).to_big_endian(&mut output);
-		log::trace!(target: "staking-precompile", "🥩output is {:?}", output);
-
-		let gas_consumed = <Runtime as pallet_evm::Config>::GasWeightMapping::weight_to_gas(
-			<Runtime as frame_system::Config>::DbWeight::get().read,
-		);
+		// Fetch info.
+		gasometer.record_cost(RuntimeHelper::<Runtime>::db_read_gas_cost())?;
+		let is_selected = parachain_staking::Pallet::<Runtime>::is_selected_candidate(&address);
 
+		// Build output.
 		Ok(PrecompileOutput {
 			exit_status: ExitSucceed::Returned,
-			cost: gas_consumed,
-			output: output.to_vec(),
-			logs: Default::default(),
+			cost: gasometer.used_gas(),
+			output: EvmDataWriter::new().write(is_selected).build(),
+			logs: vec![],
 		})
 	}
 
-	// The dispatchable wrappers are next. They return a substrate inner Call ready for dispatch.
-
-	fn join_candidates(input: &[u8]) -> Result<parachain_staking::Call<Runtime>, ExitError> {
-		let amount = parse_amount::<BalanceOf<Runtime>>(&input[..32])?;
-		let collator_candidate_count = parse_weight_hint(&input[32..])?;
+	// Runtime Methods (dispatchables)
 
-		log::trace!(target: "staking-precompile", "Collator stake amount is {:?}", amount);
-		log::trace!(
-			target: "staking-precompile",
-			"Weight Hint: collator candidate count is {:?}",
-			collator_candidate_count
-		);
+	fn join_candidates(
+		mut input: EvmDataReader,
+		context: &Context,
+	) -> Result<
+		(
+			<Runtime::Call as Dispatchable>::Origin,
+			parachain_staking::Call<Runtime>,
+		),
+		ExitError,
+	> {
+		// Read input.
+		input.expect_arguments(2)?;
+		let raw_amount = input.read::<u128>()?;
+		let amount: BalanceOf<Runtime> = raw_amount
+			.try_into()
+			.map_err(|_| error("balance type conversion failed"))?;
```

Comment:
This is also verbose. Let me know if there is a less verbose way of reading balance type using the existing code @librelois or @nanocryk . If not, I'll add `read_balance<Balance: TryFrom<U256>>` back, which I originally wrote in [this commit in 632](ttps://github.com/PureStake/moonbeam/pull/632/commits/be8679d358c57177a9c111c3d00a6066e3697fac)

Reply:
There is more than one instance of this code which is why I'm proposing something less verbose.

Reply:
It's not optimal to use a U256 when we use a u128 as concrete type for the balance.
In fact you just have to declare as constraint `Runtime::Balance: EvmData`, it will work because all unsigned integers implement EvmData. And it will allow you to write in one line: 

```
let amount: BalanceOf<Runtime> = input.read()?;
```

Reply:
Thank you @librelois !